### PR TITLE
Improving ID-CNN tagger accuracy

### PR DESCRIPTION
**Pull request by shira-g**

### CODE REVIEW

File: nlp_architect/models/tagging.py [hunk 1]
```diff
@@ -495,7 +529,7 @@ def evaluate_predictions(self, logits, label_ids):
             logits = logits.detach().cpu().numpy()
         out_label_ids = active_labels.detach().cpu().numpy()
         y_true, y_pred = self.extract_labels(out_label_ids, logits)
-        p, r, f1 = tagging(y_pred, y_true)
+        p, r, f1 = tagging(y_pred, y_true, self.bilou_format)
```
Comment (peteriz) [inline nlp_architect/models/tagging.py]: can you make `tagging` be label type agnostic?

File: nlp_architect/procedures/token_tagging.py [hunk 2]
```diff
@@ -143,6 +143,14 @@ def add_parse_args(parser: argparse.ArgumentParser):
                         help="Overwrite the content of the output directory")
     parser.add_argument("--no_cuda", action='store_true',
                         help="Avoid using CUDA when available")
+    parser.add_argument('--log_file', type=str, default="log_file",
```
Comment (peteriz) [inline nlp_architect/procedures/token_tagging.py]: please remove

File: nlp_architect/procedures/token_tagging.py [hunk 3]
```diff
@@ -143,6 +143,14 @@ def add_parse_args(parser: argparse.ArgumentParser):
...
+                            help='log path for evaluation output')
+    parser.add_argument("--bilou", action='store_true',
+                        help="whether to use bilou format evaluation")
+    parser.add_argument('--drop_penalty', type=float, default=1e-4,
+                        help='idcnn dropout penalty')
+    parser.add_argument('--ignore_token', type=str, default="",
+                        help='a token to ignore when processing the data')
```
Comment (peteriz) [inline nlp_architect/procedures/token_tagging.py]: are these new arguments supported by all the taggers?

File: nlp_architect/utils/embedding.py [hunk 4]
```diff
@@ -100,13 +100,15 @@ def get_embedding_matrix(embeddings: dict, vocab: Vocabulary,
     else:
         mat = np.zeros((len(vocab), emb_size))
     for word, wid in vocab.vocab.items():
-        vec = embeddings.get(word.lower(), None)
+        vec = embeddings.get(word, None)
+        if vec is None:
+            vec = embeddings.get(word.lower(), None)
```
Comment (peteriz) [inline nlp_architect/utils/embedding.py]: some word embedding models are lower case only
this method should be modified to support selection of lower/mixed mode

File: nlp_architect/utils/metrics.py [hunk 5]
```diff
@@ -92,8 +92,8 @@ def pearson_and_spearman(preds, labels):
...
 
 
-def tagging(preds, labels):
-    p = precision_score(labels, preds)
-    r = recall_score(labels, preds)
-    f1 = f1_score(labels, preds)
+def tagging(preds, labels, bilou_format = False):
+    p = precision_score(labels, preds, bilou_format)
+    r = recall_score(labels, preds, bilou_format)
+    f1 = f1_score(labels, preds, bilou_format)
```
Comment (peteriz) [inline nlp_architect/utils/metrics.py]: should be format agnostic
or split into 2 different metrics

File: nlp_architect/models/tagging.py [hunk 6]
```diff
@@ -51,7 +52,7 @@ class NeuralTagger(TrainableModel):
     """
 
     def __init__(self, embedder_model, word_vocab: Vocabulary, labels: List[str] = None,
-                 use_crf: bool = False, device: str = 'cpu', n_gpus=0):
+                 use_crf: bool = False, device: str = 'cpu', n_gpus=0, bilou_format = False):
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: because it affects the evaluation method

File: nlp_architect/models/tagging.py [hunk 7]
```diff
@@ -64,8 +65,21 @@ def __init__(self, embedder_model, word_vocab: Vocabulary, labels: List[str] = N
             self.crf = CRF(self.num_labels, batch_first=True)
         self.device = device
         self.n_gpus = n_gpus
+        self.bilou_format = bilou_format
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: same

File: nlp_architect/models/tagging.py [hunk 8]
```diff
@@ -64,8 +65,21 @@ def __init__(self, embedder_model, word_vocab: Vocabulary, labels: List[str] = N
...
+        if all(c.isupper() for c in string):
+            return 1 #"AA"
+        if string[0].isupper():
+            return 2 #"Aa"
+        if any(c for c in string if c.isupper()):
+            return 3 #"aAa"
+        else:
+            return 4 #"a"
+
+
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: done

File: nlp_architect/models/tagging.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: removed

File: nlp_architect/models/tagging.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: removed

File: nlp_architect/models/tagging.py [hunk 9]
```diff
@@ -238,6 +271,12 @@ def train(self, train_data_set: DataLoader,
                 if self.n_gpus > 1:
                     loss = loss.mean()
 
+                # add dropout penalty
+                logits_no_drop = self.model(**inputs, no_dropout=True)
+                sub = logits.sub(logits_no_drop)
+                drop_loss = torch.div(torch.sum(torch.pow(sub,2)) , 2)
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: this method is used in the paper

File: nlp_architect/models/tagging.py (hunk 1 above)
Comment (shira-g) [reply nlp_architect/models/tagging.py]: done

File: nlp_architect/utils/metrics.py (hunk 5 above)
Comment (shira-g) [reply nlp_architect/utils/metrics.py]: done

File: nlp_architect/procedures/token_tagging.py (hunk 3 above)
Comment (shira-g) [reply nlp_architect/procedures/token_tagging.py]: yes

File: nlp_architect/utils/embedding.py (hunk 4 above)
Comment (shira-g) [reply nlp_architect/utils/embedding.py]: done

File: nlp_architect/models/tagging.py [hunk 10]
```diff
@@ -187,7 +208,9 @@ def train(self, train_data_set: DataLoader,
               logging_steps: int = 50,
               save_steps: int = 100,
               save_path: str = None,
-              distiller: TeacherStudentDistill = None):
+              distiller: TeacherStudentDistill = None,
+              log_file: str = None,
```
[the remark this answers is not in the document]
Comment (shira-g) [reply nlp_architect/models/tagging.py]: changed arg to 'best_result_file'

File: nlp_architect/procedures/token_tagging.py (hunk 2 above)
Comment (shira-g) [reply nlp_architect/procedures/token_tagging.py]: changed arg to 'best_result_file'

File: nlp_architect/data/utils.py [hunk 11]
```diff
@@ -94,9 +94,13 @@ def read_tsv(input_file, quotechar=None):
...
 
 
-def read_column_tagged_file(filename: str, tag_col: int = -1):
+def read_column_tagged_file(filename: str, tag_col: int = -1, ignore_line: str = None):
     """Reads column tagged (CONLL) style file (tab separated and token per line)
     tag_col is the column number to use as tag of the token (defualts to the last in line)
+    Args:
+        filename (str): input file path
+        tag_col (int): the column contains the labels
+        ignore_line (str): a str token to exclude
```
Comment (peteriz) [inline nlp_architect/data/utils.py:103]: ignore line? or ignore a token? please revise naming

File: nlp_architect/models/tagging.py [hunk 12]
```diff
@@ -202,13 +219,16 @@ def train(
         save_steps: int = 100,
         save_path: str = None,
         distiller: TeacherStudentDistill = None,
+        best_result_file: str = None,
```
Comment (peteriz) [inline nlp_architect/models/tagging.py:222]: why store the best result here? this should be stored as part of the training script

File: nlp_architect/models/tagging.py [hunk 13]
```diff
@@ -269,145 +356,28 @@ def train(
...
-                test = self._get_eval(test_data_set, "test")
-                if dev > best_dev:
-                    best_dev = dev
-                    best_test = test
-                    if save_path is not None and save_best:
-                        self.save_model(save_path)
-                logger.info("Best result: dev= %s, test= %s", str(best_dev), str(best_test))
-            if save_path is not None and global_step % save_steps == 0:
-                self.save_model(save_path)
+        self.update_best_model(
```
Comment (peteriz) [inline nlp_architect/models/tagging.py:371]: again, `self.update_best_model` shouldn't be a part of the model but as part of the training

File: nlp_architect/models/transformers/base_model.py [hunk 14]
```diff
@@ -281,6 +281,8 @@ def _train(
         max_grad_norm: float = 1.0,
         logging_steps: int = 50,
         save_steps: int = 100,
+        training_args: str = None,
+        best_result_file: str = None,
```
Comment (peteriz) [inline nlp_architect/models/transformers/base_model.py:285]: should not be here
if you want to save the best model this should be configured when building the model

File: nlp_architect/models/transformers/base_model.py [hunk 15]
```diff
@@ -281,6 +281,8 @@ def _train(
         max_grad_norm: float = 1.0,
         logging_steps: int = 50,
         save_steps: int = 100,
+        training_args: str = None,
```
Comment (peteriz) [inline nlp_architect/models/transformers/base_model.py:284]: why do you provide more training args?  this should have been set in the model or the training script

File: nlp_architect/models/transformers/token_classification.py [hunk 16]
```diff
@@ -306,6 +305,8 @@ def train(
         max_grad_norm: float = 1.0,
         logging_steps: int = 50,
         save_steps: int = 100,
+        training_args: bool = None,
```
Comment (peteriz) [inline nlp_architect/models/transformers/token_classification.py:308]: same comment as before

File: nlp_architect/utils/metrics.py [hunk 17]
```diff
@@ -94,3 +122,409 @@ def tagging(preds, labels):
...
+    if prev_tag == "O" and tag == "I":
+        chunk_start = True
+
+    if tag != "O" and tag != "." and prev_type != type_:
+        chunk_start = True
+
+    return chunk_start
+
+
+def f1_score(y_true, y_pred, suffix=False):
```
Comment (peteriz) [inline nlp_architect/utils/metrics.py:263]: Please rename entity metrics to reflect that these are NER metrics (f1 and accuracy exist for other tasks and it might be confusing)

File: nlp_architect/data/utils.py (hunk 11 above)
Comment (shira-g) [reply nlp_architect/data/utils.py:103]: done

File: nlp_architect/utils/metrics.py (hunk 17 above)
Comment (shira-g) [reply nlp_architect/utils/metrics.py:263]: done

File: nlp_architect/models/transformers/base_model.py (hunk 15 above)
Comment (shira-g) [reply nlp_architect/models/transformers/base_model.py:284]: moved to model

File: nlp_architect/models/transformers/token_classification.py (hunk 16 above)
Comment (shira-g) [reply nlp_architect/models/transformers/token_classification.py:308]: moved to model
